irohad/ametsuchi/impl/rocksdb_common.hpp

```diff
@@ -0,0 +1,1229 @@
+  // ➡️ permissions
+  static auto constexpr kGranted{FMT_STRING(
+      RDB_PATH_ACCOUNT /**/ RDB_GRANTABLE_PER /**/ RDB_XXX /**/ RDB_XXX)};
+
+  // key ➡️ value
+  static auto constexpr kSetting{
+      FMT_STRING(RDB_ROOT /**/ RDB_WSV /**/ RDB_SETTINGS /**/ RDB_XXX)};
+
+  /**
+   * ######################################
+   * ############## FILES #################
+   * ######################################
+   */
+  // domain_id ➡️ default role
+  static auto constexpr kDomain{FMT_STRING(RDB_PATH_DOMAIN)};
+
+  // domain_id/account_name
+  static auto constexpr kQuorum{
+      FMT_STRING(RDB_PATH_ACCOUNT /**/ RDB_OPTIONS /**/ RDB_F_QUORUM)};
+
+  // account_domain_id/account_name ➡️ size
+  static auto constexpr kAccountAssetSize{
+      FMT_STRING(RDB_PATH_ACCOUNT /**/ RDB_OPTIONS /**/ RDB_F_ASSET_SIZE)};
+}  // namespace iroha::ametsuchi::fmtstrings
+
+#undef RDB_ADDRESS
+#undef RDB_TLS
+#undef RDB_OPTIONS
+#undef RDB_F_ASSET_SIZE
+#undef RDB_PATH_DOMAIN
+#undef RDB_PATH_ACCOUNT
+#undef RDB_F_QUORUM
+#undef RDB_DELIMITER
+#undef RDB_ROOT
+#undef RDB_STORE
+#undef RDB_WSV
+#undef RDB_NETWORK
+#undef RDB_SETTINGS
+#undef RDB_ASSETS
+#undef RDB_ROLES
+#undef RDB_TRANSACTIONS
+#undef RDB_ACCOUNTS
+#undef RDB_PEERS
+#undef RDB_STATUSES
+#undef RDB_DETAILS
+#undef RDB_GRANTABLE_PER
+#undef RDB_POSITION
+#undef RDB_TIMESTAMP
+#undef RDB_DOMAIN
+#undef RDB_SIGNATORIES
+#undef RDB_ITEM
+
+namespace {
+  auto constexpr kValue{FMT_STRING("{}")};
+}
+
+namespace iroha::ametsuchi {
+
+  /**
+   * RocksDB transaction context.
+   */
+  struct RocksDBContext {
+    /// RocksDB transaction
+    std::unique_ptr<rocksdb::Transaction> transaction;
+
+    /// Buffer for key data
+    fmt::memory_buffer key_buffer;
+
+    /// Buffer for value data
+    std::string value_buffer;
+  };
+
+  /**
+   * Port to provide access to RocksDB instance.
+   */
+  struct RocksDBPort {
+    RocksDBPort(RocksDBPort const &) = delete;
+    RocksDBPort &operator=(RocksDBPort const &) = delete;
+    RocksDBPort() = default;
+
+    void initialize(std::string const &db_name) {
+      rocksdb::Options options;
+      options.create_if_missing = true;
+      options.error_if_exists = true;
+
+      rocksdb::OptimisticTransactionDB *transaction_db;
+      auto status = rocksdb::OptimisticTransactionDB::Open(
+          options, db_name, &transaction_db);
+
+      std::unique_ptr<rocksdb::OptimisticTransactionDB> tdb(transaction_db);
+      if (!status.ok())
+        throw std::runtime_error(status.ToString());
+
+      transaction_db_.swap(tdb);
+    }
+
+    void prepareTransaction(RocksDBContext &tx_context) {
+      assert(transaction_db_);
+      tx_context.transaction.reset(
+          transaction_db_->BeginTransaction(rocksdb::WriteOptions()));
+      tx_context.key_buffer.clear();
+      tx_context.value_buffer.clear();
+    }
+
+   private:
+    std::unique_ptr<rocksdb::OptimisticTransactionDB> transaction_db_;
+  };
+
+  class IrohaDbError : public std::runtime_error {
```

Comment:
Since it's already in `iroha` namespace, having `Iroha` in name seems redundant. There is also a convention to return error codes in usual execution paths. Exceptions seem more appropriate for initialization, or cases when something "unexpected" happens, and an alternative execution path has to be taken. In case of command errors, it's not an unexpected case when the client creates a stateless valid command which will be invalid in stateful validation.

Reply:
We have no performance bottleneck in exception usage. But using them here makes code much more clear and easy instead of error-forwarding. I tried both of the variants.